cpp/src/hdbscan/detail/condense.cuh

```diff
@@ -18,17 +18,154 @@
 
 #include <cub/cub.cuh>
 #include <thrust/copy.h>
+#include <thrust/count.h>
 #include <thrust/device_ptr.h>
 #include <thrust/execution_policy.h>
 #include <thrust/extrema.h>
 #include <thrust/fill.h>
 #include <thrust/reduce.h>
 
+#include <omp.h>
+
+#include <algorithm>
+#include <cstdint>
+
 namespace ML {
 namespace HDBSCAN {
 namespace detail {
 namespace Condense {
 
+template <typename value_idx, typename value_t>
+value_t get_lambda_h(value_idx node, value_idx num_points, const value_t* deltas)
+{
+  value_t delta = deltas[node - num_points];
+  return delta > 0.0 ? 1.0 / delta : std::numeric_limits<value_t>::max();
+}
+
+/*
+ * Bottom-up implementation using omp threads
+ * This is called from build_condensed_hierarchy
+ */
+template <typename value_idx, typename value_t>
+void condense_hierarchy_cpu(const std::vector<value_idx>& children,
```

Comment:
We don't usually distinguish between `gpu` and` cpu` versions following this semantic. I'd rename this to `_condense_hierarchy_bottom_up` or similar to indicate the algorithmic difference here. Whether this runs on the CPU or GPU is a secondary concern.